### DISCUSSION TITLE
Extend BaseAuthStore model with event hooks

### DISCUSSION THREAD

**Discussion by charbs-io** [pocketbase/pocketbase#3081] [Q&A]
Is there a way to use event hooks (the experimental JS hooks) to modify the [BaseAuthStore model](https://github.com/pocketbase/js-sdk#authstore)?

I am looking to make a user's workspace details available with the authStore model, without having to fetch any collections on the UI. (For context, workspace is a collection that is referenced by another collection workspaceUsers (which itself uses the user.id and the workspace.id to keep track of which user has access to which workspace(s))).

I see the [onrecordauthrequest](https://pocketbase.io/docs/js-event-hooks/#onrecordauthrequest) docs and I have a feeling that might be what I want, but the docs are unclear on how i can _modify_ the response...

Thanks folks!

**Comment by ganigeorgiev** [comment]
I think there is some misunderstanding or at least I don't understand what you are trying to do.
The JS SDK AuthStore is not related to the [JS app hook](https://pocketbase.io/docs/js-overview/). The JS app hooks run only on the server together with your executable.

> but the docs are unclear on how i can modify the response

Modifying the event fields (`e.record`, `e.token`, `e.meta`) would affect the final auth response.

If you want a completely different response format, you can eager write a new response body with `e.httpContext.json(200, {"something": 123})` (_see [JS Writing Response body](https://pocketbase.io/docs/js-routing/#writing-response-body)_).

Note that if you just want to `expand` an auth record relation in the auth response, you can do something like:

```js
// note: with the planned JS SDK refactoring the empty object will not be required in the future
await pb.collection("users").authWithPassword("test@example.com", "1234567890", {}, { expand: "someRelField" });
```

**Reply by charbs-io** [reply]
Thanks @ganigeorgiev . I'm aware the JS SDK AuthStore is not related to the app hooks. I am looking to modify the event fields for the auth response, so that when I use the JS SDK to log the user in, i have a different response (one that has more data than default).

I think your response has enough context and should steer me in the right direction... Will report back!

**Reply by charbs-io** [reply]
@ganigeorgiev - exactly what i was looking for. 

```
e.httpContext.json(200, {
      record: e.record,
      token: e.token,
      meta: {
        workspaces: [
          {
            id: "123",
            name: "example",
          },
        ],
      },
    });
```

I can fetch my workspaces and attach them to the response! This is great because it saves the client side another round trip to get extra data after successful auth.

Thanks

**Reply by ganigeorgiev** [reply]
_Please keep in mind that the above will overwrite the original `meta` fields in case OAuth2 auth is used. Also note that the `meta` field are usually not stored as part of the JS SDK AuthStore `model`._

**Reply by charbs-io** [reply]
Thanks for letting me know. For those looking in the future, i attached what i needed on the e.record.

Something like this
```
const recordPayload = {
    ...e.record,
    workspaces: {...},
  };

  e.httpContext.json(200, {
    record: recordPayload,
    token: e.token,
  });
```

**Reply by ganigeorgiev** [reply]
@cayoub88 The above note about the `meta` field still remains valid if you are writing the response body manually with `e.httpContext.json`. To avoid it, you can directly modify ~`e.record = recordPayload`~(1). The hook finalizer will read the event fields and will write the response for you.

(1) this actually may not work in your case because e.record no longer will be a `Record` model. Instead you can do the following:

```js
e.record.set("workspaces", workspaces);
e.record.withUnknownData(true) // allow exporting non-schema fields
```

**Reply by charbs-io** [reply]
Thanks @ganigeorgiev that's a better approach (using `set`). It works for my usecase.

However, I would have expected that `workspaces` would remain a part of the authstore model when i retrieve it later from the backend.

```
// /pb_hooks/somefile.pb.ts

routerAdd(
  "POST",
  "/api/extend/workspace/switch/:workspaceId",
  (c) => {
    const workspaceId = c.pathParam("workspaceId");

    $apis.requestInfo(c);

    const info = $apis.requestInfo(c);
    const record = info;

    return c.json(200, {
      message: "Hello " + workspaceId,
      record: record
    });
  }
);
```

The response `record.authRecord` does not have the additional `workspaces` array I `set` earlier. I guess this was my original question, is there a way to do set additional props on authStore and have them available for the remainder of the session?

**Reply by ganigeorgiev** [reply]
> I would have expected that workspaces would remain a part of the authstore model...

But why?

> ... is there a way to do set additional props on authStore and have them available for the remainder of the session?

No, the AuthStore updates only when an auth action is performed or when the SDK can be certain that the updated record is part of the AuthStore (eg. `update()`). For every other case (including custom routes) you'll have to update it on your own using `pb.authStore.save(token, model)`.

Note that there are no sessions. Each request has its own state and updating the auth record on the backend will not update the authStore. You can subscribe to realtime changes of your auth record using `collection("users").subscribe(pb.authStore.model.id, ...)`, but this will get triggered only when there is a db action on the related user record. Adding a dynamic/fake property on the fly to the record will not trigger it.

**Reply by charbs-io** [reply]
To answer the "but why" question @ganigeorgiev  - it's to return a list of workspaces that the user belongs to + keep track of which workspace the user is currently "active" in.

I was trying this approach because it meant I could keep my workspace and workspaceUsers collections seperate from my user collection, and not have to worry about updating data twice (and I would get this workspace data for free with the login call on the UI). 

I suppose I can change the approach by updating the user collection to include this data (perhaps a simple array of relations to workspace or workspaceUsers), but that will require manual updates to the user collection whenever they are added to a new workspace (or removed from one). But that will still work and achieve what I need. I guess I am trying to push pocketbase to its limits here 😃

**Reply by ganigeorgiev** [reply]
I personally would keep the collections separate and subscribe to those specific collections if I wanted to listen for changes.

The optimization gains (if any) by sending the "workspace" with the auth record would probably be negligible in your case compared to a separate http request (_and in some situations the latter could result in faster loading times for the end users because the separate read requests can be processed concurrently_).

**Reply by charbs-io** [reply]
Appreciate your input, valuable as always!
